.github/workflows/ci.yml

```diff
@@ -0,0 +1,98 @@
+name: ci
+
+on:
```

Comment:
Some thoughts on the CI configs for the various branches  - please share your thoughts on this:
- PR to all branches except master: lint, build
- PR to master: lint, build, **and test**
- (optional) push to master, develop, or staging: lint, build

### Lint and build PRs to all branches
We want to ensure that PRs to all branches have basic code hygiene. This is especially so for PRs to develop and master branch, but we can also imagine that developers make PRs from feat-branch-A-continued into feat-branch-A-base for multi-part PRs.

### Also lint and build on push to master, develop, or staging branches
These are actually optional, since Netlify does the build and deploy for us; but no harm running them again on GitHub Actions.

### Only run e2e test on PRs to master
The Cypress e2e test has a lot of side effects (resets the git history of the `e2e-test-repo`, the entire e2e test suite cannot be run concurrently by two different CI runners), so we want to use it sparingly.

It is quite tempting to use the e2e test in many places - but really, the e2e test should be mainly used as a final release test to ensure that our release is functional and doesn't break any existing features.

The main reason why we are so tempted to run the e2e test in other places, e.g. PR to develop, is because we don't have other integration/unit tests at the moment. So that's something that we can work on next.

Reply:
We could implement the above by either:
- trying out the GitHub Actions [on.workflow_run.<branches|branches_ignore>](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_runbranchesbranches-ignore)
- a gatekeep method as used by [some of our other products](https://github.com/opengovsg/checkfirst/blob/develop/.github/workflows/ci.yml#L71-L86) - it feels very clunky to me, so the first is preferable

Reply:
**deciding workflow**
Actually i think i got muddled with this PR - the points you've raised are excellent but an additional distinction could be made between e2e/non-e2e tests. 

This distinction is necessary to decide what tests to run on which branch and i think this whole workflow (the e2e tests) workflow should only be run on PRs against `master` ie, releases. This is because this workflow is both expensive but realistic, which means that it should only be used sparingly in high stakes PRs.

conversely, against every other branch/push, what could happen is we do the workflow that you mentioned with the addition of `npm test`. this command doesn't do anything right now but is a guarantee that when (not if!) we add unit/integration tests in the future, we will get safety on pushes!

 **implementing workflow**
i agree that the gatekeep method is clunky! i think we can just have 2 workflow files using the `branches|branches_ignore` pattern and it should work. a BIG caveat here is that in the period between now till writing non-e2e tests, we lose safety on snyk/major version pushes on dependabot. this is non-ideal because for dependabot, major version pushes have breaking changes. i tried to mitigate this originally with the `push` event + running e2e but i think a suitable alternative might be to just run tests manually for the branch or to allow `develop` as a base and run only if the head ref matches a regex.